### PR TITLE
Introduced timeouts for MSAL calls.

### PR DESCRIPTION
**Pull request by machavan**
Description:

Calls to MSAL authentication library done from the Java driver code use future.get(), potentially leading to hangs in case the library does not receive timely response from the server. All these calls done as part of getFedAuthToken need to have timeouts imposed. The timeout period is calculated based on remaining time wrt login timeout, and passed to future.get(<timeout>, <timeunit>) calls with this fix.

Tests:
- Existing unit tests pass on local build
- All ADO tests passed
- Added new tests

### CODE REVIEW

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java [hunk 1]
```diff
@@ -1321,4 +1323,34 @@ public void testServerNameField() throws SQLException {
...
+
+    @Test
+    public void testGetSqlFedAuthTokenFailure() throws SQLException {
+        try (Connection conn = getConnection()){
+        	SqlFedAuthInfo fedAuthInfo = ((SQLServerConnection) conn).new SqlFedAuthInfo();
+        	fedAuthInfo.spn = "https://database.windows.net/";
+        	fedAuthInfo.stsurl = "https://login.windows.net/xxx";
+        	SqlAuthenticationToken fedAuthToken = SQLServerMSAL4JUtils.getSqlFedAuthToken(fedAuthInfo, "xxx",
+                    "xxx",SqlAuthentication.ACTIVE_DIRECTORY_PASSWORD.toString(), 0);
+        	fail("Expected the test to throw SQLServerException");
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1336]: fail(TestResource.getResource("R_expectedExceptionNotThrown"));

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java [hunk 2]
```diff
@@ -1321,4 +1323,34 @@ public void testServerNameField() throws SQLException {
...
+        } catch (SQLServerException e) {
+        	//test pass
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1338]: check that the error message is the one we expect?

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java [hunk 3]
```diff
@@ -1321,4 +1323,34 @@ public void testServerNameField() throws SQLException {
...
+
+    @Test
+    public void testGetSqlFedAuthTokenPrincipalFailure() throws SQLException {
+        try (Connection conn = getConnection()){
+        	SqlFedAuthInfo fedAuthInfo = ((SQLServerConnection) conn).new SqlFedAuthInfo();
+        	fedAuthInfo.spn = "https://database.windows.net/";
+        	fedAuthInfo.stsurl = "https://login.windows.net/xxx";
+        	SqlAuthenticationToken fedAuthToken = SQLServerMSAL4JUtils.getSqlFedAuthTokenPrincipal(fedAuthInfo, "xxx",
+                    "xxx",SqlAuthentication.ACTIVE_DIRECTORY_SERVICE_PRINCIPAL.toString(), 0);
+        	fail("Expected the test to throw SQLServerException");
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1351]: fail(TestResource.getResource("R_expectedExceptionNotThrown"))

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java [hunk 4]
```diff
@@ -1321,4 +1323,34 @@ public void testServerNameField() throws SQLException {
...
+        } catch (SQLServerException e) {
+    	}
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1353]: check error msg

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 5]
```diff
@@ -297,7 +302,7 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipalCertificate(SqlFedAuthI
 
             final CompletableFuture<IAuthenticationResult> future = clientApplication
                     .acquireToken(ClientCredentialParameters.builder(scopes).build());
-            final IAuthenticationResult authenticationResult = future.get();
+            final IAuthenticationResult authenticationResult = future.get(millisecondsRemaining, TimeUnit.MILLISECONDS);
```
Comment (Gueorgi) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:306]: Why there is no below the same condition as above:
`} catch (TimeoutException e) {
            throw new SQLServerException(SQLServerException.getErrString("R_connectionTimedOut"), e);
        }`

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 6]
```diff
@@ -6110,10 +6110,11 @@ private SqlAuthenticationToken getFedAuthToken(SqlFedAuthInfo fedAuthInfo) throw
         }
 
         while (true) {
+            int millisecondsRemaining = timerRemaining(timerExpire);
```
Comment (Gueorgi) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:6113]: If this is 0 or negative does it make sense to try the calls below but not throw timed out exception here (or from timerRemaining call)? Or just let the first calls throw the exception?

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java (hunk 1 above)
Comment (Gueorgi) [reply src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1336]: Also needs one test with negative time maybe

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java (hunk 6 above)
Comment (JJ-WM) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:6113]: there's min value of 1ms, but the question is still valid.

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java (hunk 2 above)
Comment (machavan) [reply src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1338]: Done

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java (hunk 6 above)
Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:6113]: right, we are just letting the first call throw it

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java (hunk 4 above)
Comment (machavan) [reply src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java]: done

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java (hunk 3 above)
Comment (machavan) [reply src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java]: done

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java (hunk 1 above)
Comment (machavan) [reply src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java]: done

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 7]
```diff
@@ -151,10 +159,12 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipal(SqlFedAuthInfo fedAuth
...
         Set<String> scopes = new HashSet<>();
         scopes.add(scope);
-
-        lock.lock();
-
+        
+        boolean lockAcquired = false;
         try {
+        	//Just try to acquire the lock and if can't then proceed to attempt to get the token
+            lockAcquired = lock.tryLock(Math.min(millisecondsRemaining, TOKEN_LOCK_WAIT_DURATION_MS), TimeUnit.MILLISECONDS);
```
Comment (001mertiya) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:166]: Where are we using `lockAcquired`? We should go ahead only if it returns true(lock acquired), right?

Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:166]: The idea is to attempt acquiring the lock with a timeout of 5 seconds. If the lock cannot be acquired within this time, proceed to attempt token acquisition regardless of whether the earlier thread(s) have released the lock or not

`lockAcquired` is used to check in the finally block if the lock should be released or not.

Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:166]: The purpose is to optimize the token acquisition process, the first caller succeeding does caching which is then leveraged by subsequent threads.  However, if the first thread takes considerable time, then we want the others to also go and try after waiting for a while.

If we were to let say 30 threads try in parallel, they would all miss the cache and hit the AAD auth endpoints to get their tokens at the same time, stressing the auth endpoint.

Comment (JJ-WM) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:166]: Understood the intent, but would Semaphore be a more readable solution?

Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java]: Agree. The code has been updated along with detailed comment describing the intent.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 8]
```diff
@@ -77,19 +80,28 @@ private SQLServerMSAL4JUtils() {
...
+            String authenticationString, int millisecondsRemaining) throws SQLServerException {
         ExecutorService executorService = Executors.newSingleThreadExecutor();
 
         if (logger.isLoggable(Level.FINEST)) {
             logger.finest(LOGCONTEXT + authenticationString + ": get FedAuth token for user: " + user);
         }
 
-        lock.lock();
-
+        boolean semAcquired = false;
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:93]: coding standard - boolean should be named isXXX ie isSemAcquired

Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java]: Done

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 9]
```diff
@@ -219,9 +248,18 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipalCertificate(SqlFedAuthI
         Set<String> scopes = new HashSet<>();
         scopes.add(scope);
 
-        lock.lock();
-
+        boolean isSemAcquired = false;
         try {
+            //
+            //Just try to acquire the semaphore and if can't then proceed to attempt to get the token.
```
Comment (divang) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:254]: Can we intergate mockito framework and simulate this scenario?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 10]
```diff
@@ -116,7 +128,7 @@ static SqlAuthenticationToken getSqlFedAuthToken(SqlFedAuthInfo fedAuthInfo, Str
                     .builder(Collections.singleton(fedAuthInfo.spn + SLASH_DEFAULT), user, password.toCharArray())
                     .build());
 
-            final IAuthenticationResult authenticationResult = future.get();
+            final IAuthenticationResult authenticationResult = future.get(Math.min(millisecondsRemaining, TOKEN_WAIT_DURATION_MS), TimeUnit.MILLISECONDS);
```
Comment (saurabh500) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:131]: Why is this min computation needed? In case of infinite connection timeout ?

Comment (saurabh500) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:131]: Also in case of infinite timeout, what is the value of millisecondsRemaining? Is it 0 or negative?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 11]
```diff
@@ -77,19 +80,28 @@ private SQLServerMSAL4JUtils() {
...
         try {
+            //
+            //Just try to acquire the semaphore and if can't then proceed to attempt to get the token.
+            //The purpose is to optimize the token acquisition process, the first caller succeeding does caching 
+            //which is then leveraged by subsequent threads. However, if the first thread takes considerable time, 
+            //then we want the others to also go and try after waiting for a while.
+            //If we were to let say 30 threads try in parallel, they would all miss the cache and hit the AAD auth endpoints 
+            //to get their tokens at the same time, stressing the auth endpoint.
+            //
+            isSemAcquired = sem.tryAcquire(Math.min(millisecondsRemaining, TOKEN_SEM_WAIT_DURATION_MS), TimeUnit.MILLISECONDS);
```
Comment (saurabh500) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:103]: Based on the comment, it feels like there should be a `if` condition check to see if the semaphore actually was acquired or not. 
If not acquired, then it means that someone else is holding the semaphore and take actions ? 
From the current code, whether the semaphore was acquired or not, would lead to next code to execute no ? 
If the semaphore was not acquired, then we may have waited long enough, and that should likely lead to a timeout of connection attempt ?

Comment (saurabh500) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:103]: This pattern is used at a lot of places in the code. 
So same comment is applicable at all usage instance.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 12]
```diff
@@ -340,10 +382,19 @@ static SqlAuthenticationToken getSqlFedAuthTokenIntegrated(SqlFedAuthInfo fedAut
...
+
+        	final PublicClientApplication pca = PublicClientApplication
```
Comment (saurabh500) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:397]: Nit: The indentation of this line seems off by one tab. Or is it Github UI messing this up for me ? :D

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java [hunk 13]
```diff
@@ -1321,4 +1323,51 @@ public void testServerNameField() throws SQLException {
...
+    }
+
+    @Test
+    public void testGetSqlFedAuthTokenFailureNoWaiting() throws SQLException {
+        try (Connection conn = getConnection()){
+        	SqlFedAuthInfo fedAuthInfo = ((SQLServerConnection) conn).new SqlFedAuthInfo();
+        	fedAuthInfo.spn = "https://database.windows.net/";
+        	fedAuthInfo.stsurl = "https://login.windows.net/xxx";
+        	SqlAuthenticationToken fedAuthToken = SQLServerMSAL4JUtils.getSqlFedAuthToken(fedAuthInfo, "xxx",
+                    "xxx",SqlAuthentication.ACTIVE_DIRECTORY_PASSWORD.toString(), 0);
```
Comment (saurabh500) [inline src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1350]: Out of curiosity, time remaining of 0, does it mean no more time left, or does it mean infinite?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java (hunk 10 above)
Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:131]: millisecondsRemaining is computed based on loginTimeout and TOKEN_WAIT_DURATION_MS is set to 20 seconds based on previous experience that anything over 15 seconds is not likely to return a token at all and we should bail out and retry in the outer connection retry loop.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java (hunk 11 above)
Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:103]: The idea is to block for a bit to allow the first request to get a token then use the cached token. But if a request hangs, we don’t want to block everyone for very long. It’s not going to break anything if the semaphore isn’t acquired. It’s just means two callers might overwrite the same token cache entry.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java (hunk 12 above)
Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java]: Done

File: src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java (hunk 13 above)
Comment (machavan) [reply src/test/java/com/microsoft/sqlserver/jdbc/SQLServerConnectionTest.java:1350]: Right 0 means no time left here for the purpose of this test

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java (hunk 9 above)
Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:254]: We have covered test scenarios in unit tests by directly calling methods with specified millisecondsRemaining as arguments. 

But we will surely want to add Mockito based tests as well to try out few more scenarios.

**Review by saurabh500 — APPROVED** [review]

**Review by divang — APPROVED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java [hunk 14]
```diff
@@ -77,19 +80,28 @@ private SQLServerMSAL4JUtils() {
         throw new UnsupportedOperationException(SQLServerException.getErrString("R_notSupported"));
     }
 
-    private static final Lock lock = new ReentrantLock();
+    private static final Semaphore sem = new Semaphore(1);
```
Comment (imasud00) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: >Semaphore [](http://example.com/codeflow?start=45&length=9)

If I have more than 1 connections tring with different auth mode should we restrict this semaphore to 1. Unless this class has its own instance for each connectionstring. ?

Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: The lock is replaced by semaphore here only to improve the readability of the code as requested in some code review comments here, since it is not really serving a mutual exclusion but rather a kind of timed that helps all the incoming requests not to bombard AAD server at once and also not to wait for a long time for a previous requester to finish (if it is taking a long time).

But this is an interesting point to consider if there should be a separate semaphore per Auth mode.

Comment (David-Engel) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: We could technically partition the locking/semaphore to be by auth endpoint (fedAuthInfo.stsurl), but for the purposes of simplicity we've done it globally. The primary use case is connection pools that open a lot of connections to the same server (and auth point) at the same time. Different pools/threads opening many simultaneous connections to different servers (and auth points) isn't a use case we have prioritized here. Let me know if you disagree.

Comment (imasud00) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: actually we should partition becasue if one token is expired we need to wait all others connecting string as well .
 Is my understanding correct ?

Comment (David-Engel) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: A small potential wait in that scenario, yes.

Partitioning will add yet another lock, though. Obtain lock1 to add a new TokenCacheMap entry that contains a lock for the entry. Obtain the TokenCacheMap entry lock. Release lock1. Proceed with token acquisition. Release TokenCacheMap entry lock. Don't forget to refactor the cache expiry code to take into account the new lock...

Comment (imasud00) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: how does this work when we are working with Integrated Login where each user token is going to be different.

Comment (imasud00) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: Create an instance of this class per connections string could solve unmercenary coupling amongst connection string.

However I am still curious on how does this work if a webapp is using integrated login or integrated login, where each token will be unique to the user.

Comment (machavan) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java:83]: Created an ADO story to explore this further. 

Resolving this discussion here.

**Review by divang — DISMISSED** [review]

**Review by David-Engel — APPROVED** [review]
